### PR TITLE
#7 - Solution for Collections.emptyList vs Collections.EMPTY_LIST

### PR DESCRIPTION
**Pull request by michal-majewski**
I have prepared some code with a brief description about obtaining an empty list from the Collection class. Let me know if it is an acceptable way to introduce this topic.

### CODE REVIEW

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+         * */
+        List<Integer> listFromEmptyListMethodWithIntegerType = Collections.emptyList();
+    }
+
+    private static void checkIfListIsImmutable(final List emptyList, final String msgForSuccessfulModification, final String msgForUnsuccessfulModification) {
+        try {
+            emptyList.add(1);
+            System.out.println(msgForSuccessfulModification);
+        } catch (UnsupportedOperationException e) {
+            System.out.println(msgForUnsuccessfulModification);
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:115]: Why not `System.err.println`?

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+     * Using this method is likely to have comparable cost to using the like-named field.
+     * (Unlike this method, the field does not provide type safety.)
+     * <br>
+     * Since then it is recommended way to obtain an empty list.
+     *
+     * @see <a href="https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/Collections.html#emptyList()">Java SE 16 & JDK 16 - emptyList()</a>
+     **/
+    private <T> List<T> emptyList() {
+        return Collections.emptyList();
+    }
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:156]: This should be closer to `private final List EMPTY_LIST = Collections.EMPTY_LIST;`. in that way someone will be able to compare those two quickly.

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        checkIfListIsImmutable(Collections.EMPTY_LIST,
+                "List from the Collections.EMPTY_LIST is mutable",
+                "List from the Collections.EMPTY_LIST is immutable");
+
+        checkIfListIsImmutable(Collections.emptyList(),
+                "List from the Collections.emptyList() is mutable",
+                "List from the Collections.emptyList() is immutable");
+
+        System.out.println(Collections.EMPTY_LIST instanceof List ? "The Collections.EMPTY_LIST is an instance of List" : "The Collections.EMPTY_LIST is NOT an instance of LIst");
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:62]:
```suggestion
        System.out.println(Collections.EMPTY_LIST instanceof List ? "The Collections.EMPTY_LIST is an instance of List" : "The Collections.EMPTY_LIST is NOT an instance of List");
```

Comment (marcin-ogorzalek-EPAM) [reply src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:62]: BTW why this is not enclosed in some method as with previous calls.

Comment (marcin-ogorzalek-EPAM) [reply src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:62]: And where is javadoc?

**Review by marcin-ogorzalek-EPAM — CHANGES_REQUESTED** [review]
Some changes and answers requested.

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java (hunk 1 above)
Comment (michal-majewski) [reply src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:115]: Amendments done

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java (hunk 2 above)
Comment (michal-majewski) [reply src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:156]: Good point. Changes done

**Comment by michal-majewski** [comment]
Changes done after your remarks. I'm asking for another review.

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        checkIfListIsImmutable(Collections.emptyList(), "Collections.emptyList()");
+
+        System.out.println(checkIfProvidedEmptyListIsAnInstanceOfRawList(Collections.EMPTY_LIST)
+                ? "The Collections.EMPTY_LIST is an instance of List"
+                : "The Collections.EMPTY_LIST is NOT an instance of List");
+
+        /*
+         * The code below will cause a compilation error with the specified message:
+         * "'List' cannot be safely cast to 'List<Object>'"
+         * */
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:98]:
```suggestion
         */
```

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        //System.out.println(Collections.EMPTY_LIST instanceof List<Object>);
+
+        /*
+         * The code below is correct, and it will not cause any errors. Both conditions with "instanceof" are always true.
+         * */
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:103]:
```suggestion
         */
```

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        System.out.println(checkIfProvidedEmptyListIsAnInstanceOfRawList(Collections.emptyList())
+                ? "The Collections.emptyList() is an instance of List"
+                : "The Collections.emptyList() is not an instance of List");
+        System.out.println(checkIfEmptyListFromCollectionsStaticMethodIsAnInstanceOfObjectsList()
+                ? "The Collections.emptyList() is an instance of List<Object>"
+                : "The Collections.emptyList() is not an instance of List<Object>");
+
+        /*
+         * The return type is a raw List
+         * */
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:113]:
```suggestion
         */
```

I can see that in many places to be fixed. To end multiline comment you can just use `*/` without additional asterisk.

File: src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+        System.out.println(checkIfBothListsAreSerializable()
+                ? "Both lists are serializable"
+                : "Both lists are not serializable");
+
+        checkIfListIsImmutable(Collections.EMPTY_LIST, "Collections.EMPTY_LIST");
+
+        checkIfListIsImmutable(Collections.emptyList(), "Collections.emptyList()");
+
+        System.out.println(checkIfProvidedEmptyListIsAnInstanceOfRawList(Collections.EMPTY_LIST)
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/ess/helpers/collections/EmptyListComparison.java:91]: I would gladly see some comment here similar to one from lines 101-103.

**Review by marcin-ogorzalek-EPAM — CHANGES_REQUESTED** [review]
One additional comment required and some cosmetic changes (remember about those not pointed out directly.

Generally good job. Can be merged after fixes.

**Comment by michal-majewski** [comment]
Changes done. Third time lucky, perhaps? :)

**Review by marcin-ogorzalek-EPAM — APPROVED** [review]

**Comment by marcin-ogorzalek-EPAM** [comment]
Indeed, the third time's the charm.

**Pull request merged**